### PR TITLE
Recognize pandas CSV files with empty index labels

### LINKED ISSUE #6694: astropy.io.ascii does not read pandas csv file correctly
Pandas by default will write CSV files where the header for the first column is missing:

```python
In [4]: from pandas import DataFrame

In [5]: df = DataFrame()

In [6]: df['a'] = [1,2,3]

In [7]: df.to_csv('test.csv')

In [8]: %more test.csv
,a
0,1
1,2
2,3
```

Whether or not this is sensible is debatable, but this means there are a lot of CSV files in the wild missing the first header column name. Astropy doesn't read these in correctly though:

```
In [10]: from astropy.io.ascii import read

In [11]: read('test.csv')
Out[11]: 
<Table masked=True length=4>
 col1 col2
int64 str1
----- ----
   --    a
    0    1
    1    2
    2    3
```

I think we might want to special case this, or deal better with cases like this given how common these kinds of files are going to be.

### PR DESCRIPTION
**Pull request by drdavella**
This resolves #6694, although it is probably not the last word on the matter. It handles the case where a CSV file is generated by pandas without a label for the index column. This can be detected because the delimiter will be the first meaningful character in the header row.

The current changes only affect the fast reader, but we might need updates for the other readers as well (like the basic reader). I'm not really sure. Also, these files are now recognized without any additional parameters, but maybe we want an option to override this behavior?

### CODE REVIEW

File: CHANGES.rst [hunk 1]
```diff
@@ -25,6 +25,9 @@ astropy.extern
 astropy.io.ascii
 ^^^^^^^^^^^^^^^^
 
+- Enable processing of CSV files generated by pandas that have an empty label
+  for the index column. [#6722]
```
Comment (bsipocz) [inline CHANGES.rst]: I'm sorry to mess up your PR number...

File: astropy/io/ascii/tests/test_read.py [hunk 2]
```diff
@@ -1283,3 +1284,27 @@ def test_unsupported_read_with_encoding(tmpdir):
...
+    no_index_fname = os.path.join(str(tmpdir), 'no_index.csv')
+    df.to_csv(no_index_fname, index=False)
+
+    table = ascii.read(no_index_fname)
+    assert table.colnames == ['stuff']
+    assert all(table['stuff'] == [x for x in range(10)])
+
+    # This is the problematic case since by default the first column in a
+    # pandas csv is the index column. However, if we don't explicitly provide
+    # an index name, then the header column starts with a single column,
```
Comment (drdavella) [inline astropy/io/ascii/tests/test_read.py]: This should be "starts with a single delimiter".

File: CHANGES.rst (hunk 1 above)
Comment (drdavella) [reply CHANGES.rst]: @bsipocz it's okay, there's apparently a race condition in `next_pr_number.py` 😛

File: astropy/io/ascii/tests/test_read.py [hunk 3]
```diff
@@ -1283,3 +1284,27 @@ def test_unsupported_read_with_encoding(tmpdir):
...
         ascii.read('t/simple3.txt', guess=False, fast_reader='force',
                    encoding='latin1', format='fast_csv')
+
+
+def test_read_pandas_csv(tmpdir):
+    pandas = pytest.importorskip('pandas')
+
+    df = pandas.DataFrame()
+    df['stuff'] = [x for x in range(10)]
+    no_index_fname = os.path.join(str(tmpdir), 'no_index.csv')
```
Comment (MSeifert04) [inline astropy/io/ascii/tests/test_read.py]: can't you just use `tmpdir.join(...)` without `os`?

Comment (drdavella) [reply astropy/io/ascii/tests/test_read.py]: Ah yes, I keep forgetting about that. I thought for awhile there we were trying to maintain compatibility with older versions of pytest, but I guess that's not important anymore?

Comment (bsipocz) [reply astropy/io/ascii/tests/test_read.py]: we already require >3.1, so no need to optimize on old pytests....

File: astropy/io/ascii/tests/test_read.py [hunk 4]
```diff
@@ -8,6 +8,7 @@
 import locale
 import platform
 from io import StringIO
+import os
```
Comment (drdavella) [inline astropy/io/ascii/tests/test_read.py]: Don't need this anymore. Doh.

**Comment by drdavella** [comment]
Travis test failure is just a remote data timeout.

File: astropy/io/ascii/tests/test_read.py [hunk 5]
```diff
@@ -1283,3 +1284,27 @@ def test_unsupported_read_with_encoding(tmpdir):
...
+    assert all(table['stuff'] == [x for x in range(10)])
+
+    # This is the problematic case since by default the first column in a
+    # pandas csv is the index column. However, if we don't explicitly provide
+    # an index name, then the header column starts with a single delimiter,
+    # indicating an index header name of None
+    has_index_fname = str(tmpdir.join('has_index.csv'))
+    df.to_csv(has_index_fname)
+
+    table = ascii.read(has_index_fname)
```
Comment (taldcroft) [inline astropy/io/ascii/tests/test_read.py]: It's probably worth explicitly testing two different `format` values `basic` and `csv` in addition to the current test.  Your current test doesn't hit the `FastCsv` reader, only `FastBasic`.

One good practice when working on changes like this is always look at `ascii.get_read_trace()` after doing a read in order to see which readers were tried and (importantly) the one that was actually used for the final output.  The guess logic is complicated...

**Comment by taldcroft** [comment]
On a more general note, should the pandas index column be kept instead of just being tossed?  Maybe call the column `__index__` or `__pandas_index__` or something?

File: astropy/io/ascii/tests/test_read.py (hunk 5 above)
Comment (drdavella) [reply astropy/io/ascii/tests/test_read.py]: @taldcroft thanks for the review. Clearly this PR needs some more work, so I'll let you know when it's ready for review again. The guess logic does seem complicated, and it's hard to know how many different places this kind of change needs to be implemented and tested.

**Pull request closed without merging**